### PR TITLE
Testing code coverage

### PR DESCRIPTION
**Pull request by vasanth096**

### CODE REVIEW

**Comment by vasanth096** [comment]
Goals of this pull request
* Make sure reports are generated on every run
* Add test cases to cover missed statements
* Add test cases (inject faults) to cover statements that are triggered only during internal errors
* Automate the process of calculating the coverage and uploading the report to codacy

File: setup/docker-compose.yml [hunk 1]
```diff
@@ -31,6 +31,7 @@ services:
       - WAIT_HOSTS_TIMEOUT=120
       - PROJECT_DIR=api-server
       - JAR_NAME=api-server-fat.jar
+      - JAVA_OPTS=-javaagent:org.jacoco.agent-0.8.6-runtime.jar=destfile=target/jacoco.exec
```
Comment (pct960) [inline setup/docker-compose.yml:34]: This option needs to go in only for the test environment

File: setup/install [hunk 2]
```diff
@@ -1,8 +1,10 @@
 #!/bin/bash
 
+cd api-server
+wget https://repo1.maven.org/maven2/org/jacoco/org.jacoco.agent/0.8.6/org.jacoco.agent-0.8.6-runtime.jar
```
Comment (pct960) [inline setup/install:4]: This jar should be downloaded only when vermillion is being tested

File: tests/features/secure-files.feature [hunk 3]
```diff
@@ -71,12 +103,17 @@ Feature: VermillhFeature: Vermillion is able to handle secure files
...
 		        Then The response status should be 400
 
-	Scenario: Download secure-file with only token
-                Given Vermillion is running
-                When The consumer downloads file by passing only token
-                Then The response status should be 403
+        Scenario: Download file with public id and token
+          Given Vermillion is running
+          When The consumer downloads file by passing public id and token
+          Then The response status should be 400
```
Comment (pct960) [inline tests/features/secure-files.feature:109]: If a public file is downloaded with a token, the status should not be 400. Could you describe what is happening here?

File: tests/features/secure-timeseries.feature [hunk 4]
```diff
@@ -98,6 +104,11 @@ Feature: Vermillion is able to handle secure timeseries datasets
            When The consumer requests for a standalone authorised ID with invalid token
            Then The response status should be 403
 
+         Scenario: Authorised ID - single
```
Comment (pct960) [inline tests/features/secure-timeseries.feature:107]: The description is misleading. The title is authorised ID, but the test is checking with an unauthorised ID

File: tests/features/steps/auth_vars.py [hunk 5]
```diff
@@ -32,7 +32,7 @@
 		i] + " for 1 month"
 for i in range(6, 8):
 	acl_set_policy += "; consumer@iisc.ac.in can access example.com/test-category/" + resource_ids[
-		i] + " for 1 month if scope = read"
+		i] + " for 1 second if scope = read"
```
Comment (pct960) [inline tests/features/steps/auth_vars.py:35]: How is this being used? Is it to check for expired tokens?

File: tests/features/steps/secure-files.py [hunk 6]
```diff
@@ -14,21 +15,109 @@
...
         'file': ('sample.txt', open('sample.txt', 'rb')),
         'metadata': ('meta.json', open('meta.json', 'rb')),
     }
-    check_pub_file(params, files, context)
+    post_files(params, files, context)
+
+
+@when('The consumer downloads file by passing multiple resource ids and a token')
+def step_impl(context):
+    url = 'https://localhost/download?token=' + tokens["8_10_rw"] + '&id=' + res[8] + ',' + res[10]
```
Comment (pct960) [inline tests/features/steps/secure-files.py:75]: Use uri params in the request module. Do not hardcode this.

File: tests/features/steps/then.py [hunk 7]
```diff
@@ -86,6 +81,14 @@ def step_impl(context):
         if dat != re['hits'][0]['data']:
             raise UnexpectedBehaviourError('Secure Timeseries data not found in response')
 
+    if context.type == 'authorised_id_multiple':
+        dat = {"hello": "india"}
+        dat1 = {"hello": "world"}
+        re = context.response.json()
+        print(re)
+        if dat1 != re['hits'][0]['data']:
```
Comment (pct960) [inline tests/features/steps/then.py:89]: Better to
* check the length of the response
* and then iterate through them trying to find the desired results

Hardcoding index values is a bad idea

**Review by pct960 — CHANGES_REQUESTED** [review]

File: api-server/pom.xml [hunk 8]
```diff
@@ -159,12 +159,12 @@
         <plugin>
           <groupId>org.jacoco</groupId>
           <artifactId>jacoco-maven-plugin</artifactId>
-          <version>${jacoco.version}</version>
+          <version>0.8.6</version>
```
[the remark this answers is not in the document]
Comment (vasanth096) [reply api-server/pom.xml]: Made changes

File: setup/install (hunk 2 above)
Comment (vasanth096) [reply setup/install]: Made changes

File: tests/features/secure-files.feature (hunk 3 above)
Comment (vasanth096) [reply tests/features/secure-files.feature:109]: Only secure files can be downloaded here..

if (resourceIDStr.endsWith(".public")) {
                apiFailure(
                        context,
                        new BadRequestThrowable(
                                "This API is for secure resources only. Use /provider/public endpoint to explore public data"));
                return;
            }
        }

File: tests/features/secure-timeseries.feature (hunk 4 above)
Comment (vasanth096) [reply tests/features/secure-timeseries.feature:95]: Made changes

File: tests/features/steps/auth_vars.py (hunk 5 above)
Comment (vasanth096) [reply tests/features/steps/auth_vars.py:35]: This is to check for expired tokens to introspect the invalid ids

File: tests/features/steps/secure-files.py (hunk 6 above)
Comment (vasanth096) [reply tests/features/steps/secure-files.py:75]: made changes

File: tests/features/steps/then.py (hunk 7 above)
Comment (vasanth096) [reply tests/features/steps/then.py]: Made changes

File: setup/docker-compose.yml (hunk 1 above)
Comment (vasanth096) [reply setup/docker-compose.yml]: Made changes

**Review by pct960 — APPROVED** [review]

**Pull request merged**